### ISSUE TITLE
Stop using IntersectionObserver polyfill in ads if native support is present

### ISSUE THREAD

**Issue by cramforce** [ampproject/amphtml#5603]
Now that stable Chrome has IntersectionObserver we should use it instead of the postMessage mechanism.

CC @lannka

**Comment by lannka** [comment]
Should we just leave it to 3p creative to handle? Or you are thinking about us doing it in `integration.js`?

**Comment by cramforce** [comment]
I was thinking to do it in `integration.js`. The 3ps are unlikely to do the right thing across the board.

**Comment by lannka** [comment]
I noticed several differences between the polyfill and native impl:

1. IntersectionObserverEntry.time in the spec is the elapsed time in ms since page load. Our polyfill uses Date.time()
2. Our polyfill triggers whenever the intersection has a change, and throttled to 10/s. The native impl only triggers when the intersection ratio cross some predefined thresholds.
3. As shown in the [README](https://github.com/ampproject/amphtml/blob/master/ads/README.md#position-in-viewport), our polyfill also provides the initial intersection (on page load). However, the native impl only triggers when it has intersection. So it doesn't trigger if the iframe is BTF on page load.

> When a listener is registered, it will be called 2x in short order. Once with the position of the ad when its iframe was created and then again with the current position.

> The value window.context.initialIntersection contains the initial intersection record at the time the iframe was created.

For 1), shall we change the polyfill? Or shall we compensate to the native impl? I prefer the latter, but not sure if anyone is relying on it.
For 2), I suggest we specify a fine grained threshold array as [0. 0.05, 0.10, ..., 0.95, 1]. To make sure it gets triggered on every 5%. 
For 3), we can install the IntersectionObserver on page load (no matter a listener will be registered or not. assume this is cheap), and approximate the initial intersection to the 1st intersection record if it's received within 500ms. Otherwise, the initial intersection is considered as 0. Once a listener get registered, we will send the initial record and also the previous record to mimic the "2x in short order" behavior.

**Comment by jridgewell** [comment]
1. We're not even generating a `DOMHighResTimeStamp`. 
2. Maybe allow them to pass the thresholds they're interested in?
3. I don't understand this change.

**Comment by lannka** [comment]
@jridgewell 
1. `performance.now()` will give the value. It has a [good browser coverage](http://caniuse.com/#search=performance).
2. sure, we can allow a param passed in. To keep the backward compatibility, we can make the current behavior as default (no param given).
3. which part? the problem or the proposed solution?

**Comment by jridgewell** [comment]
1. With that, we don't need any relative stuff (`performance.now()` is already)
2. Sure, that'd simplify some of the IntersectionObserver code
3. The solution. Why install an observer if none is requested? Or anything else in it.

**Comment by lannka** [comment]
3. it's to provide the `initialIntersection`. I reconsidered this, my proposal won't work. We have to pass the info from host doc via URL fragment. 

1. the starting time is different though. in host doc, it's relative to the page load. inside iframe, it's relative to the iframe load. though most likely the creative only cares about the duration, it might still be good to consolidate. 3) requires the time to be relative to page load. That means we will intercept the InOb and subtract an offset.

**Comment by lannka** [comment]
Things become more straightforward now as we consolidated the behavior between native impl and polyfill due to @zhouyx 's work.

Assign to Yuxuan.

**Comment by zhouyx** [comment]
working on this. have two questions in advance.
1. In the case of ad iframe resize, but `amp-ad` doesn't. Native impl and our polyfill could result in very different changeEntries. Are we good with this?
2. Since we will do our test in Chrome for most cases. Do we want to disable using native impl in test mode?

**Comment by lannka** [comment]
Reopen this because native IntersectionObserver doesn't give `rootBounds` inside x-origin iframe. Some ad networks are currently relying on it:

Taboola @nitzanvolman
Teads @RonanDrouglazet
Yieldmo @yieldmo-rao 

Are you able to make changes to not relying on this value?

**Comment by yieldmo-rao** [comment]
@lannka We use `rootBounds` to get the scroll position of the top window. In normal environments, we do `window.top`, but since we do not have access to that, we use `rootBounds`. We don't mind switching it out, but what is the alternative API you would recommend to get the window's current scroll position at any given point in time?

**Comment by lannka** [comment]
@yieldmo-rao thanks for the reply. Can you further explain what do you use the scroll position for?

**Comment by yieldmo-rao** [comment]
@lannka , we use it for calculating a bunch of metrics that we report back to our server. Additionally, many of our formats are interactive (parallax effects based on scroll position), so that calculation requires us to know where we are on the screen currently.

**Comment by yieldmo-rao** [comment]
Oh, forgot to add; in addition to the "push" based model that is used with event listeners (we get updates when the user is actively scrolling), we also need to be able to "pull" the position at any given point in time (when the user may not be actively scrolling, but the format is on the screen), We only do this once or twice, but I do know that we definitely need that ability as well. With `window.scrollX` and `scrollY`, we always have that data. We did that in AMP by saving the latest intersection upon getting a scroll event.

**Comment by yieldmo-rao** [comment]
@lannka We're also going for a deployment freeze until after the new year. Owing to that, I hope that this is not something that will be changed/updated before the end of year!

**Comment by harquail** [comment]
hey @lannka is this change in production?  We're currently seeing problems with our format animation that may be related to this change.

**Comment by harquail** [comment]
Specifically it looks like we are only getting events on intersection with the viewport, rather than on scroll as with the native intersection observer (we rely on these for format positioning).  I can open a new issue once I've done more investigation.

**Comment by yieldmo-rao** [comment]
Adding on to @harquail , all our client side reporting is also broken. :(

**Comment by zhouyx** [comment]
We are not using native intersection observer now. But our polyfill may be a breaking change if you rely on intersection observer for scroll event.
we polyfill the behavior of native intersection observer and only send intersection change entry when ratio cross thresholds. (will not send on every scroll). @lannka Should we revert the change?

**Comment by harquail** [comment]
Ok that seems like the cause of our issue.  We need a way to get our actual scroll position on the page — and ideally notifications of position change.  We're open to using another API for this if available, but this is currently a breaking change for us.

**Comment by yieldmo-rao** [comment]
@zhouyx Also, any insight into when this change was rolled out? I ask so that we can go back and update our internal reports accordingly. Thank you!

**Comment by zhouyx** [comment]
@yieldmo-rao  @harquail We did introduce a change (went to prod on Dec 9th). It could be a breaking change if you rely on the frequency of the changeEntry. We used to send a maximum of 10 changeEntries per second. But now we have a threshold of [0, 0.05, 0.1, ..... 0.95, 1.0] and will only send changeEntry when ratio cross the limit. 

However to clarify we never send IntersectionObserver changeEntry out of viewport before. @harquail as you described, I don't think you can get scroll position on every scroll before. Can you investigate? Thanks

**Comment by harquail** [comment]
@zhouyx thanks for the clarification!  The breaking change for us is the change to only send changeEntry when the ratio crosses the limit.  We currently use the intersectionObserver events for format animation, which we now only get when the ad is intersecting with the viewport bounds.  We are not relying on these events when the ad is outside the viewport.

Example pages (need to emulate mobile):
http://test.yieldmo.com/test/hyperscroller.html
http://test.yieldmo.com/test/amp_hyperscroller.html

**Comment by lannka** [comment]
@harquail I suggest we expose additional param in `window.context.observeIntersection(changesCallback, opt_option)` to let you specify the threshold array. For your particular use case, you might want to have a more granular array to have smooth animation. WDYT?

**Comment by harquail** [comment]
@lannka That would solve part of our use case!  But we are also using this API currently to detect all position changes, not just those when intersecting with the viewport.  The majority of our ads animate in response to granular position changes, and we also use this information to calculate engagement (a user scrolling up and down to interact with a Yieldmo ad).

However, from amp's perspective it may not make sense to have position change in general be part of the intersectionObserver, because it does not require intersection.

**Comment by lannka** [comment]
@harquail I still don't believe our deprecated implementation worked for position changes outside viewport. 

But if this is really something you need (position change outside viewport), could you open a separate "Intent to implement" for broader discussion? As you said, IntersectionObserver is definitely not to serve that suppose.

**Comment by lannka** [comment]
@harquail I think I understand the problem now. We were sending pings 10/s when the full ad is in viewport, and you were relying on that for animation. However, now we're only sending pings on intersection ratio change (not position change).

We will consider a temporary rollback to reduce the damage at your side (@erwinmombay what's the best way to do that in this case?). As for a long term solution, we might consider introduce a different API. Relying on IntersectionObserver is kind of hacky. Could you fire an separate I2I to list your requirements for such an PositionObserver?

**Comment by erwinmombay** [comment]
@lannka if we can revert the PR that made the change cleanly without conflicts we can just cherry pick the revert and deploy that quickly

**Comment by erwinmombay** [comment]
or if we can't do a revert just try and do a undo PR that we can patch to prod

**Comment by harquail** [comment]
Thank you all!  We will make an I2I for the new API proposal.  Really appreciate your quick response.

**Comment by yieldmo-rao** [comment]
Thanks for your help everyone! REALLY appreciate this!

**Comment by zhouyx** [comment]
@jasti @lannka What's our next step on using native IntersectionObserver?

**Comment by lannka** [comment]
Once we build #6873 then we can ask everyone moves to that API.

**Comment by ampprojectbot** [comment]
This issue hasn't been  updated in awhile. @zhouyx Do you have any updates?

**Comment by zhouyx** [comment]
Blocked by #6873, deprioritize.
